packages/itwin/tree-widget/src/components/trees/stateless/external-sources-tree/ExternalSourcesTreeDefinition.ts

```diff
@@ -0,0 +1,274 @@
+      return { ...node, extendedData: { ...node.extendedData, imageId: "icon-ec-class" } };
+    }
+    return node;
+  }
+
+  public async defineHierarchyLevel(props: DefineHierarchyLevelProps) {
+    return this._impl.defineHierarchyLevel(props);
+  }
+
+  private async createRootHierarchyLevelDefinition(props: DefineRootHierarchyLevelProps): Promise<HierarchyLevelDefinition> {
+    const instanceFilterClauses = await this._selectQueryFactory.createFilterClauses({
+      filter: props.instanceFilter,
+      contentClass: { fullName: "BisCore.ExternalSource", alias: "this" },
+    });
+    return [
+      {
+        fullClassName: "BisCore.ExternalSource",
+        query: {
+          ecsql: `
+            SELECT
+              ${await this._selectQueryFactory.createSelectClause({
+                ecClassId: { selector: ECSql.createRawPropertyValueSelector("this", "ECClassId") },
+                ecInstanceId: { selector: "this.ECInstanceId" },
+                nodeLabel: {
+                  selector: await this.createCompositeLabelSelectClause({ externalSourceAlias: "this", repositoryLinkAlias: "rl" }),
+                },
+                extendedData: {
+                  imageId: "icon-document",
+                },
+                autoExpand: true,
+                supportsFiltering: true,
+              })}
+            FROM ${instanceFilterClauses.from} this
+            ${instanceFilterClauses.joins}
+            JOIN BisCore.SynchronizationConfigSpecifiesRootSources scsrs ON scsrs.TargetECInstanceId = this.ECInstanceId
+            JOIN BisCore.RepositoryLink rl ON rl.ECInstanceId = this.Repository.Id
+            ${instanceFilterClauses.where ? `WHERE ${instanceFilterClauses.where}` : ""}
+          `,
+        },
+      },
+    ];
+  }
+
+  private async createExternalSourcesGroupChildrenQuery({
+    parentNodeInstanceIds: groupIds,
+    instanceFilter,
+  }: DefineInstanceNodeChildHierarchyLevelProps): Promise<HierarchyLevelDefinition> {
+    const instanceFilterClauses = await this._selectQueryFactory.createFilterClauses({
+      filter: instanceFilter,
+      contentClass: { fullName: "BisCore.ExternalSource", alias: "this" },
+    });
+    return [
+      {
+        fullClassName: "BisCore.ExternalSource",
+        query: {
+          ecsql: `
+            SELECT
+              ${await this._selectQueryFactory.createSelectClause({
+                ecClassId: { selector: "this.ECClassId" },
+                ecInstanceId: { selector: "this.ECInstanceId" },
+                nodeLabel: {
+                  selector: await this.createCompositeLabelSelectClause({ externalSourceAlias: "this", repositoryLinkAlias: "rl" }),
+                },
+                hasChildren: {
+                  selector: `
+                    IFNULL((
+                      SELECT 1
+                      FROM BisCore.ExternalSourceAttachmentAttachesSource esaas
```

Comment:
Should be more efficient to join `bis.ExternalSourceAttachment` on its `Attaches` navigation property. Applies below too.

Reply:
Updated.